python_modules/dagster/dagster/_core/definitions/time_window_partitions.py

```diff
@@ -1704,74 +1625,267 @@ def can_deserialize(
             and data.get("num_partitions") is not None
         )
 
+    def __len__(self) -> int:
+        return self.num_partitions
+
+    def __contains__(self, partition_key: str) -> bool:
+        time_window = cast(
+            TimeWindowPartitionsDefinition, self.partitions_def
+        ).time_window_for_partition_key(partition_key)
+
+        return any(
+            time_window.start >= included_time_window.start
+            and time_window.start < included_time_window.end
+            for included_time_window in self.included_time_windows
+        )
+
+    def __eq__(self, other):
+        return (
+            isinstance(other, BaseTimeWindowPartitionsSubset)
+            and self.partitions_def == other.partitions_def
+            and self.included_time_windows == other.included_time_windows
+        )
+
+
+class TimePartitionKeyPartitionsSubset(BaseTimeWindowPartitionsSubset):
+    def __init__(
+        self,
+        partitions_def: TimeWindowPartitionsDefinition,
+        included_partition_keys: AbstractSet[str],
+    ):
+        self._partitions_def = check.inst_param(
+            partitions_def, "partitions_def", TimeWindowPartitionsDefinition
+        )
+        self._included_partition_keys = check.set_param(
+            included_partition_keys, "included_partition_keys", of_type=str
+        )
+
+    def with_partition_keys(
+        self, partition_keys: Iterable[str]
+    ) -> "BaseTimeWindowPartitionsSubset":
+        new_partitions = {*(self._included_partition_keys or []), *partition_keys}
+        return TimePartitionKeyPartitionsSubset(
+            self._partitions_def,
+            included_partition_keys=new_partitions,
+        )
+
+    @property
+    @cached_method
```

Comment:
Because it takes no arguments, you can use `@cached_property` from functools here